### ISSUE TITLE
How to use / Proxy

### ISSUE THREAD

**Issue by alt-romes** [conal/checkers#68]
Hello,

I was trying to use but as someone unfamiliar with quick check it’s quite hard to understand how to use this library, so much that I’ve decided to open this issue.

I would love to see some examples of `functor` and `monad`

From the source I understand the first argument of those functions is just to correctly infer the type of monad to test `m`, but I don’t understand the triple `(a, b, c)` and their constraints (QuickCheck related). How am I supposed to call those functions?

A thought: if the argument is discarded and is only there for matters of inference, why not use `Data.Proxy` which is what I think to be the preferred way of passing an argument for its type?

**Comment by alt-romes** [comment]
OK, thank you for the pointers, I'll get started.

I might send an example.

I have some new questions: Do you keep this API because of backwards compatibility? Don't you want to also provide a modern API? Is this package still valuable despite its less modern API because of featuring perhaps more laws?

But you say I *could* update the API to use `Proxy`?

**Comment by conal** [comment]
Hasn't `Proxy` been obsoleted by type application?

**Comment by alt-romes** [comment]
@conal Yes, mostly except for when using type families but needing the argument type still.

However, for a beginner, seeing `monad :: forall m ........ => TestBatch` it isn't clear at all how to say what kind of monad it is, while `Proxy m -> TestBatch` makes much more sense (and is better than making users call `monad` with an `undefined :: Type...`

**Comment by conal** [comment]
> when the library is about testing types it makes sense that the type is somehow "an argument" of the function

My 2¢: Type application is (GHC-)Haskell's modern, direct way of saying that the type (including higher-kinded) is an argument. The two older techniques you mention predated our ability to say so directly and so instead tried to fudge type applications with (ignored or vacuous) value applications.

**Comment by alt-romes** [comment]
@conal You're right. I still think it would depend on who we would be showing the library to (`forall m. TestBatch` does not *look* like a function that takes arguments), but good documentation could make sure it would be understood

**Comment by sjakobi** [comment]
> I have some new questions: Do you keep this API because of backwards compatibility? Don't you want to also provide a modern API? Is this package still valuable despite its less modern API because of featuring perhaps more laws?

You might be the first person to point out that the current API is bit weird with it's requirement for an arbitrary value to guide type inference. I'm not aware of any previous discussions about modernizing the API.

Backwards compatibility is a concern of course, but if we agree that a different API would be better, why stick with the old cruft?!

Regarding differences to other libraries: The main reason I went with `checkers` for the `multiset-properties` is https://github.com/andrewthad/quickcheck-classes/issues/74.

I agree with @conal though, that it would be better to use `TypeApplications` than `Proxy` these days.
